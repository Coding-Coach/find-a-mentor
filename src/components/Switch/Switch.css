.switch-container {
  position: relative;
  margin-top: 20px;
  display: flex;
  padding-left: 18px;
  justify-content: space-between;
}

.switch-container > label {
  background-color: #fff;
  text-transform: uppercase;
  color: #4a4a4a;
  font-weight: bold;
}

.switch-input {
  height: 20px;
}

.switch-input input[type=checkbox] {
  height: 0;
  width: 0;
  visibility: hidden;
}

.switch-input label {
  cursor: pointer;
  text-indent: -9999px;
  width: 60px;
  height: 30px;
  background-color: grey;
  display: block;
  border-radius: 100px;
  position: relative;
  top: -25px;
  transition: background-color 0.35s ease-in-out;
}

.switch-input label:after {
  content: '';
  position: absolute;
  top: 5px;
  left: 5px;
  width: 20px;
  height: 20px;
  background: #fff;
  border-radius: 20px;
  transition: left 0.2s ease-in-out, transform 0.2s ease-in-out;
}

.switch-input input:checked + label {
  background-color: var(--tag-color);
}

.switch-input input:checked + label:after {
  left: calc(100% - 5px);
  transform: translateX(-100%);
}
