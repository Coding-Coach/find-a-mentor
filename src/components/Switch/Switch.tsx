import { ChangeEvent, useContext } from 'react';
import styled from 'styled-components';
import { formFieldContext } from '../../Me/components/FormField/formContext';

type SwitchProps = {
  id?: string;
  label: string;
  isChecked: boolean;
  size?: 'small' | 'normal';
  onToggle(isChecked: boolean): void;
};

const SwitchContainer = styled.div`
  display: flex;
  justify-content: space-between;
`;

let SwitchInputRoot = styled.div<Pick<SwitchProps, 'size'>>`
  height: 20px;

  input[type='checkbox'] {
    height: 0;
    width: 0;
    visibility: hidden;

    &:checked + label {
      background-color: var(--tag-color);

      &:after {
        transform: translateX(-100%);

        ${props =>
          props.size === 'small'
            ? `
          left: calc(100% - 2.5px);
        `
            : `
          left: calc(100% - 5px);
        `}
      }
    }
  }

  label {
    display: block;
    cursor: pointer;
    position: relative;
    text-indent: -9999px;
    border-radius: 100px;
    background-color: grey;
    transition: background-color 0.35s ease-in-out;

    ${props =>
      props.size === 'small'
        ? `
        width: 25px;
        height: 15px;
        top: -15px;
        margin-left: 5px;
      `
        : `
        top: -25px;
        width: 60px;
        height: 30px;
      `}

    &:after {
      content: '';
      position: absolute;
      background: #fff;
      transition: left 0.2s ease-in-out, transform 0.2s ease-in-out;

      ${props =>
        props.size === 'small'
          ? `
        top: 2.5px;
        left: 2.5px;
        width: 10px;
        height: 10px;
        border-radius: 10px;
      `
          : `
        top: 5px;
        left: 5px;
        width: 20px;
        height: 20px;
        border-radius: 20px;
      `}
    }
  }
`;

const SwitchLabel = ({ id, label }: Pick<SwitchProps, 'id' | 'label'>) => (
  <label id={id} htmlFor={`switch-input-${id}`}>
    {label}
  </label>
);

const SwitchInput = ({
  onToggle,
  isChecked,
  id,
  size = 'normal',
}: Pick<SwitchProps, 'onToggle' | 'isChecked' | 'id' | 'size'>) => {
  const toggleSwitch = (event: ChangeEvent<HTMLInputElement>) => {
    onToggle(event.target?.checked);
  };

  return (
    <SwitchInputRoot size={size}>
      <input
        type="checkbox"
        id={`switch-input-${id}`}
        checked={isChecked}
        onChange={toggleSwitch}
      />
      <label id={`switch-label-${id}`} htmlFor={`switch-input-${id}`}>
        Toggle
      </label>
    </SwitchInputRoot>
  );
};

const Switch = ({
  id: idFromProps = '',
  label,
  onToggle,
  size,
  isChecked,
}: SwitchProps) => {
  const idFromContext = useContext(formFieldContext);
  const id = idFromProps || idFromContext;

  return (
    <SwitchContainer className={id}>
      <SwitchLabel id={id} label={label} />
      <SwitchInput
        isChecked={isChecked}
        size={size}
        id={id}
        onToggle={onToggle}
      />
    </SwitchContainer>
  );
};

export default Switch;
