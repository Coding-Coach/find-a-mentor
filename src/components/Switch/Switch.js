import React, { Component } from "react";

import "./Switch.css";

export default class Switch extends Component {
	state = {
		checked: false
	}

	toggleSwitch = () => {
		this.setState({
			checked: !this.state.checked	
		});
		this.props.onToggle(!this.state.checked);
	}

	render() {	
		const { id, label } = this.props;

		return (
		  <div className="switch-container">
		    <label htmlFor={id}>
		      {label}
		    </label>
		    <div className="switch-input">
			    <input type="checkbox"
		    		id={`switch-${id}`}
		    		checked={this.state.checked}
		    		onChange={this.toggleSwitch}
			    />
			    <label htmlFor={`switch-${id}`}>Toggle</label>
		    </div>
		  </div>
		);
	}
}