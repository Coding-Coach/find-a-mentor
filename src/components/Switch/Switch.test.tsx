import { render } from '@testing-library/react';
import Switch from './Switch';

describe('Switch component', () => {
  it('renders properly', () => {
    const { container } = render(
      <Switch label="label" onToggle={() => {}} isChecked={false} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call onToggle on click', () => {
    const onToggle = jest.fn();

    const { getByLabelText } = render(
      <Switch label="label" onToggle={onToggle} isChecked={false} />
    );
    getByLabelText('Toggle').click();
    expect(onToggle).toHaveBeenCalledWith(true);
  });
});
