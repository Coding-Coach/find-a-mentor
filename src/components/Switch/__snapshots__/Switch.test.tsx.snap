// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component renders properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  height: 20px;
}

.c1 input[type='checkbox'] {
  height: 0;
  width: 0;
  visibility: hidden;
}

.c1 input[type='checkbox']:checked + label {
  background-color: var(--tag-color);
}

.c1 input[type='checkbox']:checked + label:after {
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  left: calc(100% - 5px);
}

.c1 label {
  display: block;
  cursor: pointer;
  position: relative;
  text-indent: -9999px;
  border-radius: 100px;
  background-color: grey;
  -webkit-transition: background-color 0.35s ease-in-out;
  transition: background-color 0.35s ease-in-out;
  top: -25px;
  width: 60px;
  height: 30px;
}

.c1 label:after {
  content: '';
  position: absolute;
  background: #fff;
  -webkit-transition: left 0.2s ease-in-out,-webkit-transform 0.2s ease-in-out;
  -webkit-transition: left 0.2s ease-in-out,transform 0.2s ease-in-out;
  transition: left 0.2s ease-in-out,transform 0.2s ease-in-out;
  top: 5px;
  left: 5px;
  width: 20px;
  height: 20px;
  border-radius: 20px;
}

<div
  class="c0"
>
  <label
    for="switch-input-undefined"
  >
    label
  </label>
  <div
    class="c1"
  >
    <input
      id="switch-input-undefined"
      type="checkbox"
    />
    <label
      for="switch-input-undefined"
      id="switch-label-undefined"
    >
      Toggle
    </label>
  </div>
</div>
`;
